seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,227 @@
+    return json.loads(r.text)
+
+
+def list_anvil_workspaces(user, fields=None):
+    """
+    Get all the workspaces accessible by the logged-in user.
+
+    Args:
+    user (User model): who's credentials will be used to access AnVIL
+    fields (str): a comma-delimited list of values that limits the
+        response payload to include only those keys and exclude other
+        keys (e.g., to include {"workspace": {"attributes": {...}}},
+        specify "workspace.attributes").
+    """
+    session = anvil_session_store.get_session(user)
+    if fields is None:
+        r = session.get("api/workspaces")
+    else:
+        r = session.get("api/workspaces", params = {"fields": fields})
+    if r.status_code != 200:
+        raise TerraAPIException(
+            'Error: called Terra API "api/workspaces" got status: {} with a reason: {}'.format(r.status_code, r.reason))
+    return json.loads(r.text)
+
+
+def get_anvil_workspace_acl(workspace_namespace, workspace_name):
+    """
+    Request FireCloud access control list for workspace.
+
+    Args:
+        workspace_namespace (str): namespace (name of billing project) of the workspace
+        workspace_name (str): the name of the workspace
+    Returns:
+        {
+            "user1Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user2Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user3Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            }
+          }
+          :param workspace_name:
+          :param workspace_namespace:
+    """
+    if workspace_namespace and workspace_name:
+        uri = "api/workspaces/{0}/{1}/acl".format(workspace_namespace, workspace_name)
+        r = anvil_session_store.service_account_session.get(uri)
+        if r.status_code in DEFAULT_REFRESH_STATUS_CODES:  # has failed in refreshing the access code
+            anvil_session_store.service_account_session = AnvilSession(
+                service_account_info = GOOGLE_SERVICE_ACCOUNT_INFO, scopes = SOCIAL_AUTH_GOOGLE_OAUTH2_SCOPE)
+            r = anvil_session_store.service_account_session.get(uri)  # retry with the new access code
+        if r.status_code != 200:
+            raise TerraAPIException(
+                'Error: called Terra API "{}" got status: {} with a reason: {}'.format(uri, r.status_code, r.reason))
+        return json.loads(r.text)['acl']
+    else:
+        return {}
+
+
+class AnvilSessionStore(object):
+    sessions = {}
+
+    def __init__(self):
+        self._service_account_session = AnvilSession(service_account_info = GOOGLE_SERVICE_ACCOUNT_INFO,
```

Comment:
As I explain above, caching an AnVIL session can save the calls to the Google server for token validation. I prefer to keep the caching for these sessions because I need the session buffer for the service account anyway. I could work on the real cache for now if required.

I'm updating the @property for the service account session. The service account session has to be able to be re-setup if its access token and refresh token are all expired.

Reply:
I think you should use the real cache, I'm very worried about memory issues from letting this object just grow without limit as long as seqr is running

Reply:
The memory wouldn't grow much because it is a per-user session, not a per-frontend connection session. Anyway, I will look into the real cache today.

Reply:
I do not understand why you are unwilling to use a real cache as the cache

Reply:
I found that the cost of using `redis` cache for an AnVIL session only (not for data from AnVIL for now) is the same as re-creating the session from an access token each time requesting AnVIL. With the cache, there must be a session object re-construction when fetching the cache which is similar to creating a new session object.

If you are not comfortable with the memory objects for the AnVIL sessions, I could stop using AuthenticatedSession class for the AnVIL requests. I could use `requests` with a bearer header without creating any session objects. Then, I can avoid the cost of the session object including calling to Google to do token validation and the memory buffer for the session object.

Reply:
`requests` with a bearer header sounds like a way better approach than cached sessions. Is there a reason you weren't using it in the first place/ are there any downsides to this?

Reply:
The downsides include handling service account differently and expose the access tokens more frequently.

Reply:
what if for regular users we use bearer headers and then have a global initialized session just for the service account?

Reply:
It works. I will change to like this.

Reply:
The updates for regular users using bearer headers have been pushed.